library/x509.c

```diff
@@ -693,30 +701,25 @@ int mbedtls_x509_get_sig_alg( const mbedtls_x509_buf *sig_oid, const mbedtls_x50
  * be either manually updated or extensions should be parsed!)
  */
 int mbedtls_x509_get_ext( unsigned char **p, const unsigned char *end,
-                  mbedtls_x509_buf *ext, int tag )
+                          mbedtls_x509_buf *ext, int tag )
 {
     int ret;
     size_t len;
 
-    if( *p == end )
-        return( 0 );
-
-    ext->tag = **p;
-
-    if( ( ret = mbedtls_asn1_get_tag( p, end, &ext->len,
-            MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED | tag ) ) != 0 )
-        return( ret );
+    /* Extension structure use EXPLICIT tagging. That is, the actual
+     * `Extensions` structure is wrapped by a tag-length pair using
+     * the respective context-specific tag. */
+    ret = mbedtls_asn1_get_tag( p, end, &ext->len,
+              MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED | tag );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS + ret );
 
-    ext->p = *p;
-    end = *p + ext->len;
+    ext->tag = MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED | tag;
+    ext->p   = *p;
+    end      = *p + ext->len;
 
     /*
      * Extensions  ::=  SEQUENCE SIZE (1..MAX) OF Extension
-     *
-     * Extension  ::=  SEQUENCE  {
-     *      extnID      OBJECT IDENTIFIER,
-     *      critical    BOOLEAN DEFAULT FALSE,
-     *      extnValue   OCTET STRING  }
```

Comment:
Is it intentional to remove this comment explaining what "Extension" is? I read the commend as explaining what an Extension was, not that this comment was saying we parsed Extension here. Should we move the Extension documentation elsewhere or does it exist somewhere else already?

Reply:
Yes, it's fully intentional - the comment is very misleading, because it suggests that we're parsing an `Extension` structure (or its header) here, which is _not_ the case. We're parsing the two (!) headers of the `Extensions` structure: the explicit tag, followed by the `SEQUENCE OF` tag. Afterwards, `*p` points to the beginning of the header of the `Extension` structure, but the shape of the latter is entirely irrelevant to this function -- at least if I understand the code correctly.

Reply:
The comment is duplicated correctly in https://github.com/ARMmbed/mbedtls/blob/908cb42d7d3eb4322f2aa149ad5201ab7f6828c2/library/x509_crt.c#L711-L715 where it belongs.

Reply:
OK